src/linker/Linker/Annotations.cs

```diff
@@ -241,22 +256,72 @@ public bool IsRelevantToVariantCasting (TypeDefinition type)
 			return types_relevant_to_variant_casting.Contains (type);
 		}
 
-		public void Processed (IMetadataTokenProvider provider)
+		public bool SetProcessed (IMetadataTokenProvider provider)
 		{
-			processed.Add (provider);
+			if (processed.Add (provider)) {
+				Debug.Assert (marked_pending.Remove (provider));
+				return true;
+			}
+
+			return false;
 		}
 
 		public bool IsProcessed (IMetadataTokenProvider provider)
 		{
 			return processed.Contains (provider);
 		}
 
+		public IEnumerable<TypeDefinition> PendingPreserve ()
+		{
+			foreach (var type in pending_preserve.ToArray ())
+				yield return type;
+		}
+
+		public bool SetAppliedPreserve (TypeDefinition type, TypePreserve preserve)
+		{
+			if (!preserved_types.TryGetValue (type, out (TypePreserve preserve, bool applied) existing)) {
+				throw new InvalidOperationException ();
```

Comment:
This should throw `InternalErrorException` and include description of what went wrong so that we get at least some idea what to look for. (similarly below)

Reply:
Happy to change it to `InternalErrorException` - but I don't really understand why we would use that instead of the more descriptive existing .NET exception types - what's the intended use of one vs the other?

Reply:
It's basically similar to ICE in compilers. I think it's a good differentiation between the linker itself throwing an exception and getting an exception from some underlying framework functionality (or Cecil). There's not much difference, I agree with that... but it feels better to be consistent (the fact that we don't do this everywhere is another matter).